In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\salom\OneDrive\Documentos\11-2025 Instacart Basket Analysis'

In [3]:
df_custs = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
df_custs.shape

(206209, 10)

In [5]:
df_custs.head(10)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

In [6]:
df_custs.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [7]:
df_custs.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

#### The count of user_id matches the count of values in other columns, indicating that each row is unique. The age range is between 18 and 81 years, with a mean of 49, which seems within a reasonable range. The number of dependants and the income range seem reasonable values as well.

In [8]:
# Count values for object columns
df_custs.value_counts('STATE')

STATE
Alabama                 4044
Alaska                  4044
Arizona                 4044
Arkansas                4044
California              4044
Colorado                4044
Connecticut             4044
Delaware                4044
District of Columbia    4044
Florida                 4044
Georgia                 4044
Hawaii                  4044
Idaho                   4044
Illinois                4044
Indiana                 4044
Iowa                    4044
Kansas                  4043
Kentucky                4043
Louisiana               4043
Maine                   4043
Maryland                4043
Massachusetts           4043
Michigan                4043
Minnesota               4043
Mississippi             4043
Missouri                4043
Montana                 4043
Nebraska                4043
Nevada                  4043
New Hampshire           4043
New Jersey              4043
New Mexico              4043
New York                4043
North Carolina          4043
North Da

#### Customers are evenly distributed across states.

In [9]:
df_custs.value_counts('Gender')

Gender
Male      104067
Female    102142
Name: count, dtype: int64

#### There are slightly more male than female customers.

In [10]:
df_custs.value_counts('fam_status')

fam_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

#### There are significantly more married customers than any other family status.

In [11]:
df_custs.value_counts('date_joined')

date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

In [12]:
# Count missing values

df_custs.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [13]:
# View missing values in a subset

df_nan = df_custs[df_custs['First Name'].isnull() == True]

In [14]:
df_nan

user_id First Name    Surnam  Gender         STATE  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

#### Because the missing first name does not affect our analysis, these rows should not be removed. An 'Unknown' value can be used to impute the null values.

In [15]:
# Impute missing values with 'Unknown'

df_custs['First Name'] = df_custs['First Name'].fillna('Unknown')

In [16]:
df_custs.head(50)

user_id   First Name     Surnam  Gender           STATE  Age date_joined  \
0     26711      Deborah   Esquivel  Female        Missouri   48    1/1/2017   
1     33890     Patricia       Hart  Female      New Mexico   36    1/1/2017   
2     65803      Kenneth     Farley    Male           Idaho   35    1/1/2017   
3    125935     Michelle      Hicks  Female            Iowa   40    1/1/2017   
4    130797          Ann    Gilmore  Female        Maryland   26    1/1/2017   
5    133128      Cynthia      Noble  Female        Kentucky   43    1/1/2017   
6    152052        Chris     Walton    Male         Montana   20    1/1/2017   
7    168851       Joseph    Hickman    Male  South Carolina   30    1/1/2017   
8     69965       Jeremy       Vang    Male           Texas   47    1/1/2017   
9     82820        Shawn      Chung    Male        Virginia   26    1/1/2017   
10    26441       Gloria   Stafford  Female          Nevada   19    1/1/2017   
11    33835        Roger      Casey    Male        Nebraska   45    1/1/2017   
12    75032      Cynthia   Mcdaniel  Female         Georgia   75    1/1/2017   
13       51    Stephanie     Jensen  Female         Wyoming   37    1/1/2017   
14      516        Peter       Hunt    Male        Colorado   51    1/1/2017   
15     6965      Brandon       Mata    Male    North Dakota   70    1/1/2017   
16     8794          Amy       Mack  Female        Nebraska   30    1/1/2017   
17     9944        Brian     Oneill    Male       Wisconsin   26    1/1/2017   
18    13313       Victor       Moss    Male          Alaska   31    1/1/2017   
19    17335      Patrick     Fowler    Male         Vermont   65    1/1/2017   
20    29864        Randy      Mills    Male    North Dakota   39    1/1/2017   
21    37744        Wanda      Salas  Female        Arkansas   77    1/1/2017   
22    50000       Sandra      Trejo  Female           Maine   64    1/1/2017   
23    55567       Martha      Hurst  Female  North Carolina   67    1/1/2017   
24    57549        Shawn      Moore    Male         Montana   72    1/1/2017   
25    60386        Wayne   Woodward    Male          Alaska   59    1/1/2017   
26    68032      Charles  Stevenson    Male   West Virginia   74    1/1/2017   
27    83910         Ruth       Rice  Female         Indiana   70    1/1/2017   
28    86936     Kimberly     Monroe  Female          Oregon   57    1/1/2017   
29    99199     Kimberly     Monroe  Female        Arkansas   80    1/1/2017   
30   100123        Steve     Rogers    Male         Florida   31    1/1/2017   
31   104359         Carl     Benton    Male           Idaho   66    1/1/2017   
32   112664     Nicholas      Logan    Male      California   64    1/1/2017   
33   113616      Stephen    Murillo    Male    Pennsylvania   57    1/1/2017   
34   117740         Lisa     Sparks  Female          Oregon   55    1/1/2017   
35   127020         Judy    Barrett  Female            Ohio   35    1/1/2017   
36   135871        Doris      Mckee  Female     Connecticut   80    1/1/2017   
37   139692      Barbara      Chang  Female         Arizona   54    1/1/2017   
38   149755         Adam     Brewer    Male       Louisiana   72    1/1/2017   
39   150670        Terry  Wilkerson    Male            Iowa   55    1/1/2017   
40   155598        Marie      Beard  Female      Washington   73    1/1/2017   
41   159323      Rebecca       Tran  Female       Wisconsin   75    1/1/2017   
42   165553      Lillian   Schaefer  Female     Connecticut   71    1/1/2017   
43   166899      Kathryn   Guerrero  Female        New York   56    1/1/2017   
44   168047      Shirley   Harrison  Female          Alaska   37    1/1/2017   
45   180190      Douglas     Hebert    Male     Connecticut   30    1/1/2017   
46   180333  Christopher      Munoz    Male      Washington   80    1/1/2017   
47   195665      Brandon   Mckinney    Male    North Dakota   57    1/1/2017   
48   197419        Craig      Clark    Male   West Virginia   36    1/1/2017   
49  

In [17]:
df_custs.isnull().sum()

user_id         0
First Name      0
Surnam          0
Gender          0
STATE           0
Age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [18]:
# Check for mixed-type data

for col in df_custs.columns.tolist():
  weird = (df_custs[[col]].map(type) != df_custs[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_custs[weird]) > 0:
    print (col)

#### No column was found with mixed-type data

In [19]:
# Run a check for duplicate values

df_custs_dups = df_custs[df_custs.duplicated()]

In [20]:
df_custs_dups.shape

(0, 10)

#### There are no duplicates, as indicated by the descriptive statistics showing 206209 unique order numbers

In [21]:
# Rename columns

df_custs.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age'}, inplace = True)

In [22]:
df_custs.head(5)

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [23]:
# Changing user_id, first_name, last_name, gender, state, fam_status data type to string

df_custs['user_id'] = df_custs['user_id'].astype('str')
df_custs['first_name'] = df_custs['first_name'].astype('str')
df_custs['last_name'] = df_custs['last_name'].astype('str')
df_custs['gender'] = df_custs['gender'].astype('str')
df_custs['state'] = df_custs['state'].astype('str')
df_custs['fam_status'] = df_custs['fam_status'].astype('str')

In [24]:
df_custs.dtypes

user_id         object
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [25]:
df_custs['user_id'].dtype

dtype('O')

In [26]:
#Export clean dataframe

df_custs.to_csv(os.path.join(path, '02 Data','Prepared Data', 'customers_clean.csv'))

## Combine your customer data with the rest of your prepared Instacart data.

In [27]:
# Import last pkl dataframe

ords_prods = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_aggregated.pkl'))

In [28]:
ords_prods.value_counts('user_id')

user_id
201268    3704
129928    3637
164055    3061
186704    2936
176478    2921
          ... 
188345       3
49488        3
124615       2
203875       2
91567        1
Name: count, Length: 206209, dtype: int64

#### Count of unique customers coincides for both dataframes.

In [29]:
# Changing user_id data type in aggregated dataframe to string

ords_prods['user_id'] = ords_prods['user_id'].astype('str')

In [30]:
ords_prods['user_id'].dtype

dtype('O')

In [31]:
# Drop unnecessary columns for the merged dataframe. 
# Since every user has a unique ID, first and last names are not necessary for the analysis. 
# Date joined is also not necessary for the analysis of combined dataframes.

df_custs_m = df_custs.drop(columns = ['first_name', 'last_name', 'date_joined'])


In [32]:
df_custs_m.head()

user_id  gender       state  age  n_dependants fam_status  income
0   26711  Female    Missouri   48             3    married  165665
1   33890  Female  New Mexico   36             0     single   59285
2   65803    Male       Idaho   35             2    married   99568
3  125935  Female        Iowa   40             0     single   42049
4  130797  Female    Maryland   26             1    married   40374

In [33]:
# Merge dataframes 

ords_prods_custs = ords_prods.merge(df_custs_m, on = 'user_id', indicator = True)

In [34]:
ords_prods_custs.head(10)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2539329       1             1                   2                  8   
2   2539329       1             1                   2                  8   
3   2539329       1             1                   2                  8   
4   2539329       1             1                   2                  8   
5   2398795       1             2                   3                  7   
6   2398795       1             2                   3                  7   
7   2398795       1             2                   3                  7   
8   2398795       1             2                   3                  7   
9   2398795       1             2                   3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  ...  spending_flag  \
0                                     Soda  ...    Low spender   
1  Organic Unsweetened Vanilla Almond Milk  ...    Low spender   
2                      Original Beef Jerky  ...    Low spender   
3               Aged White Cheddar Popcorn  ...    Low spender   
4         XL Pick-A-Size Paper Towel Rolls  ...    Low spender   
5                                     Soda  ...    Low spender   
6                               Pistachios  ...    Low spender   
7                      Original Beef Jerky  ...    Low spender   
8                   Bag of Organic Bananas  ...    Low spender   
9               Aged White Cheddar Popcorn  ...    Low spender   

   avg_days_prior_order         frequency_flag  gender    state age  \
0             20.259259  Non-frequent customer  Female  Alabama  31   
1             20.259259  Non-frequent customer  Female  Alabama  31   
2             20.259259  Non-frequent customer  Female  Alabama  31   
3             20.259259  Non-frequent customer  Female  Alabama  31   
4             20.259259  Non-frequent customer  Female  Alabama  31   
5             20.259259  Non-frequent customer  Female  Alabama  31   
6             20.259259  Non-frequent customer  Female  Alabama  31   
7             20.259259  Non-frequent customer  Female  Alabama  31   
8             20.259259  Non-frequent customer  Female  Alabama  31   
9             20.259259  Non-frequent customer  Female  Alabama  31   

  n_dependants  fam_status income  _merge  
0            3     married  40423    both  
1            3     married  40423    both  
2            3     married  40423    both  
3            3     married  40423    both  
4            3     married  40423    both  
5            3     married  40423    both  
6            3     married  40423    both  
7            3     married  40423    both  
8            3     married  40423    both  
9            3     married  40423    both  

[10 rows x 30 columns]

In [35]:
# Remove unwanted columns

ords_prods_custs = ords_prods_custs.drop(columns = ['_merge'])

In [36]:
ords_prods_custs.shape

(32404859, 29)

In [37]:
ords_prods_custs.describe()

order_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.710745e+06  1.714230e+01        2.738867e+00       1.342515e+01   
std    9.872988e+05  1.753532e+01        2.090077e+00       4.246380e+00   
min    2.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    8.559470e+05  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.711049e+06  1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.565499e+06  2.400000e+01        5.000000e+00       1.600000e+01   
max    3.421083e+06  9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_prior_order    product_id  add_to_cart_order     reordered  \
count            3.032876e+07  3.240486e+07       3.240486e+07  3.240486e+07   
mean             1.110408e+01  2.559866e+04       8.352547e+00  5.895873e-01   
std              8.779064e+00  1.408400e+04       7.127071e+00  4.919087e-01   
min              0.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00   
25%              5.000000e+00  1.354400e+04       3.000000e+00  0.000000e+00   
50%              8.000000e+00  2.530200e+04       6.000000e+00  1.000000e+00   
75%              1.500000e+01  3.794700e+04       1.100000e+01  1.000000e+00   
max              3.000000e+01  4.968800e+04       1.450000e+02  1.000000e+00   

           aisle_id  department_id        prices     max_order  \
count  3.240486e+07   3.240486e+07  3.240486e+07  3.240486e+07   
mean   7.119612e+01   9.919792e+00  7.790885e+00  3.305217e+01   
std    3.821139e+01   6.281485e+00  4.241483e+00  2.515525e+01   
min    1.000000e+00   1.000000e+00  1.000000e+00  1.000000e+00   
25%    3.100000e+01   4.000000e+00  4.200000e+00  1.300000e+01   
50%    8.300000e+01   9.000000e+00  7.400000e+00  2.600000e+01   
75%    1.070000e+02   1.600000e+01  1.130000e+01  4.700000e+01   
max    1.340000e+02   2.100000e+01  2.500000e+01  9.900000e+01   

       avg_price_by_user  avg_days_prior_order           age  n_dependants  \
count       3.240486e+07          3.240485e+07  3.240486e+07  3.240486e+07   
mean        7.790885e+00          1.137498e+01  4.946527e+01  1.501896e+00   
std         7.347146e-01          6.056176e+00  1.848558e+01  1.118865e+00   
min         1.000000e+00          0.000000e+00  1.800000e+01  0.000000e+00   
25%         7.378478e+00          6.852941e+00  3.300000e+01  1.000000e+00   
50%         7.811883e+00          9.884956e+00  4.900000e+01  2.000000e+00   
75%         8.229228e+00          1.486500e+01  6.500000e+01  3.000000e+00   
max         2.320000e+01          3.000000e+01  8.100000e+01  3.000000e+00   

             income  
count  3.240486e+07  
mean   9.943773e+04  
std    4.305727e+04  
min    2.590300e+04  
25%    6.700400e+04  
50%    9.661800e+04  
75%    1.279120e+05  
max    5.939010e+05

In [38]:
# Export data to pkl

ords_prods_custs.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers.pkl'))